### PR TITLE
Add support for Xcode 9

### PR DESCRIPTION
**Pull request by tgoyne**
Compiles but doesn't actually work yet (the Float <-> NSNumber conversions crash even when building as Swift 3.2).

### CODE REVIEW

**Comment by tgoyne** [comment]
Now successfully builds and passes tests with Xcode 8.3 and Xcode 9 in both swift 3 and swift 4 modes, so I guess it's ready for review.

**Comment by tgoyne** [comment]
Going to file a radar about the `NSNumber as! Float`.

**Comment by austinzheng** [comment]
It looks like some of the CI builds (tvOS and watchOS) are failing to compile with errors, while other builds are just causing the compiler to segfault.

**Comment by jpsim** [comment]
Should we standardize on `@objc` rather than `@objc dynamic`?

**Comment by tgoyne** [comment]
`@objc` without `dynamic` doesn't work.

**Comment by jpsim** [comment]
That's very good to know (and surprising)

**Comment by austinzheng** [comment]
My impression was that `dynamic` prevents the compiler from allowing the method to be invoked by anything other than dynamic dispatch, while `@objc` doesn't make that same guarantee if dynamic dispatch isn't required (e.g. if the method is not actually ever invoked from Objective-C code).

**Comment by rkim-homma** [comment]
awesome pull request. Thanks for rolling it in.

**Comment by tgoyne** [comment]
Remaining CI failures appear unrelated (test failure in the sync progress notifier and failing to do anything at all in the device test).

**Comment by cnstoll** [comment]
Any update on this? Should we build from TG's branch for Xcode 9?

**Comment by marcuswu0814** [comment]
@cnstoll, I build branch by self using cocoapod and it work.

**Comment by kashifshaikh** [comment]
Does that mean with Swift 4 we're going to have to add both @objc along with dynamic to all our properties?

Yuck.

**Comment by tgoyne** [comment]
Yes, with Swift 4, exposing things to the obj-c runtime has been decoupled from disabling static dispatch and we need both parts.

Property behaviors will *hopefully* make it possible for us to switch to a pure-Swift solution and not need `@objc` or `dynamic`, but those aren't coming in Swift 4.

File: RealmSwift/ThreadSafeReference.swift [hunk 1]
```diff
@@ -84,7 +83,6 @@ public class ThreadSafeReference<Confined: ThreadConfined> {
     public init(to threadConfined: Confined) {
         let bridged = (threadConfined as! AssistedObjectiveCBridgeable).bridged
         swiftMetadata = bridged.metadata
-        type = type(of: threadConfined)
```
Comment (ddevjani) [inline RealmSwift/ThreadSafeReference.swift]: how to resolve error

**Pull request merged**